### PR TITLE
scbuild: rewrite

### PR DESCRIPTION
**Pull request by postamar**
This commit rewrites the scbuild package allowing it to grow gracefully
over time as we add support for new statements.

Notably, this commit moves all statement planning logic, which turns DDL
statement AST to schema change nodes, into a new internal scbuildstmts
internal package.

This commit introduces a scbuildstmts.BuildCtx interface to expose
a variety of convenience methods. This interface is deliberately
implemented in the scbuild package, for ease of testing, but also in an
effort to prevent this interface to become a god-object like the planner
in sql.

This BuildCtx interface builds on the Dependencies, BuilderState and
EventLogState interfaces:
- Dependencies remains unchanged and contains the external dependencies
  required by the scbuild and scbuildstmts packages.
- BuilderState exposes the operations on planned nodes and all mutable
  state related to eventlog support.
- EventLogState exposes the operations for updating the current metadata
  which decorates the targets added to the BuilderState.

The code in scbuildimpl deliberately follows a stylistic pattern which
isn't necessary but hopefully will be imitated as the package grows,
thereby making it easier to navigate despite it inevitably growing
quite large.

Release note: None

### CODE REVIEW

**Comment by postamar** [comment]
This PR came to be because the current state of scbuild annoyed me enough to want to do something about it. I made these changes with long-term evolution in mind. Of course the current structure I'm proposing may not be the right one, still, it should be easier to move to a new structure from an existing structure than from a big ball of mud.

Review guide: don't bother looking at the diffs in `scbuild/...`, instead check out this branch, then have a look at `scbuild`, then `scbuildutils`, and finally `scbuildimpl`. The code in there was moved about and doesn't look the same but it should do the same thing and not be much more incorrect than before, according to the tests.

**Comment by ajwerner** [comment]
I like the motivation behind this change and I like some of what it's doing, but I have a some negative gut reactions to some of the patterns.

For one, I'm not sure I get the principles are dividing what belongs in `scbuild.Builder` as opposed to `scbuildutils.BuildContext`. 

1) My gut is that it'd be cool if we could define a big ol' interface that the `scbuildimpl` will consume or a collection of them which you compose. It's fine even if there's only really one impl, because it'll let us inject mocks to write a whole bunch of unit tests. It's some up-front toil, but I feel like it'll be worth it.

Other things that I feel weird about:

2) The scoped `MakeWithNewSourceElementID` signature. I get why you did it, but it feels really opaque. If we did 1) above then we could have a method which returns a new context from the old context which corresponds to a new ID. 

3) The `StateBuilder` constructor returning a struct and then a callback to extract some state from that struct. I'd rather just a method on the struct. We can hide that method from the `impl` via the interface used by `impl` not exposing it.

**Comment by postamar** [comment]
Thanks for sharing your gut reactions this is exactly what I'm looking for right now, I've gone deep down a rabbit-hole and can't see the big picture right now.

1. I'm down with that. I hadn't even thought about the usefulness of unit tests here. So, basically another layer of DI? Why not. 
2. I got carried away with closures, I gues. Interfaces > closures. Sounds good.
3. Agreed. I did that because I didn't want to offer the temptation of messing with the output in `scbuildimpl`, but really, having that method return a cloned state should be fine in that regard. I can even add a comment there explaining why I think it's bad.

Worth mentioning that I kept ALTER TABLE ... ADD COLUMN much as it was but there's a bunch of things in there that don't feel right, like fishing out the add-primary-index element to append a stored column, yikes. I'm leaving that be until we give ADD COLUMN the proper treatment it deserves.

**Comment by postamar** [comment]
I made changes following our offline chat. It wasn't quite as straightforward as expected. This is ready for another look now.

File: pkg/sql/schemachanger/scbuild/scbuildctx/errors.go
Comment (ajwerner) [inline pkg/sql/schemachanger/scbuild/scbuildctx/errors.go:23]: I'd be down to move the errors out to either the root `schemachanger` package or an `scerrors` package. What do you think? Then, at that point, I think we could remove any imports of `scbuildctx`

**Review by ajwerner — COMMENTED** [review]
The object pattern here I like. At first I wasn't sure, but the more I digest it, the more I'm into it.

However, I still am not getting the `scbuild`/`scbuildctx` package split. As far as I can tell, other than one tiny table descriptor helper and some errors defs, the only thing `scbuildimpl` imports from `scbuildctx` are interfaces. Perhaps you're thinking that it's going to be gnarly to have other folks importing symbols like `scbuildimpl.Dependencies`. Well, I agree, so, let's use the neat declaration forwarding trick to pull them back down into scbuild. 

Concretely, how would you feel about:

1) Defining the interfaces used by `scbuildimpl` in `scbuildimpl`. 
2) Forwarding the relevant external interface (namely `Dependencies`) using:

```go
package scbuild

type (
    Dependencies = scbuildimpl.Dependencies
    CatalogReader = scbuildimpl.CatalogReader
    AuthorizationAccessor = scbuildimpl.AuthorizationAccessor
    ResolveParams = scbuildimpl.ResolveParams
)
```
3) Combining the logic in `scbuild` and `scbuildctx` into `scbuild` except the tiny bits which are directly invoked by `scbuildimpl` which you can move lower in the hierarchy because they don't depend on anything. 

Lastly, 🚲 🏠 , how do you feel about renaming `scbuildimpl` to `scbuildstmts`?

**Comment by postamar** [comment]
Thanks for the review.

>  let's use the neat declaration forwarding trick to pull them back down into scbuild.

Oh wow, TIL! Thanks for this! This Is The Way.

File: pkg/sql/schemachanger/scbuild/build.go
Comment (postamar) [reply pkg/sql/schemachanger/scbuild/build.go:16]: Bad proofreading on my part, sorry. Thanks for noticing.

File: pkg/sql/schemachanger/scbuild/scbuildctx/builder_state.go
Comment (postamar) [reply pkg/sql/schemachanger/scbuild/scbuildctx/builder_state.go:45]: Again, bad proofreading. Will fix.

File: pkg/sql/schemachanger/scbuild/build.go
Comment (postamar) [reply pkg/sql/schemachanger/scbuild/build.go:120]: I'll nix it. Generally I'm not super enthusiastic about adding stuff that doesn't get used in the first place, not sure why I got carried away here.

File: pkg/sql/schemachanger/scbuild/scbuildctx/errors.go
Comment (postamar) [reply pkg/sql/schemachanger/scbuild/scbuildctx/errors.go:91]: :galaxybrain:

File: pkg/sql/schemachanger/scbuild/scbuildctx/errors.go
Comment (postamar) [reply pkg/sql/schemachanger/scbuild/scbuildctx/errors.go:23]: Great idea.

File: pkg/sql/schemachanger/scbuild/scbuildctx/builder_state.go
Comment (postamar) [reply pkg/sql/schemachanger/scbuild/scbuildctx/builder_state.go:45]: Actually this one is still used and I'm reluctant to remove it.

**Comment by postamar** [comment]
I followed your valuable suggestions, tightened `BuilderState` by removing its embedded `Dependencies`, and the result pleases me and doesn't feel slightly weird anymore. This is worth looking at again now.

File: pkg/sql/schemachanger/scbuild/build.go
Comment (ajwerner) [inline pkg/sql/schemachanger/scbuild/build.go:58]: why define this here and not in `scbuildstmts`? I don't see a big motivation for exporting each individual statement from that package. If you did that, I'd just call this `Process` and maybe as an extra nit, lose the final `s` in that package name so this entrypoint would become `scbuildstmt.Process(ctx, bCtx, stmt)`

File: pkg/sql/schemachanger/scbuild/build.go
Comment (ajwerner) [inline pkg/sql/schemachanger/scbuild/build.go:39]: now I'm being supremely nit-picky, but aesthetically, how do you feel about:

```go
bs := newBuilderState(dependencies, initial, n)
if err := scbuildstmt.Process(ctx, makeBuildCtx(deps, bs), stmt); err != nil {
    return scpb.State{}, err
}
return *bs.output, nil
```

File: pkg/sql/schemachanger/scbuild/build.go
Comment (ajwerner) [inline pkg/sql/schemachanger/scbuild/build.go:110]: nit: call this `makeBuilderState` or return a pointer. In practice, this thing does end up being mutable. I think I'd return a pointer.

File: pkg/sql/schemachanger/scbuild/build.go
Comment (ajwerner) [reply pkg/sql/schemachanger/scbuild/build.go:39]: Another thought that occurs to me is that we can make `BuildCtx` embed `context.Context` and then eliminate an argument all over the place. It is a context after all. Then this would look like:

```go
if err := scbuildstmt.Process(makeBuildCtx(ctx, deps, bs), stmt); err != nil {
    return scpb.State{}, err
}
```
where
```go
func makeBuildCtx(ctx context.Context, deps Dependencies, builderState BuilderState) buildCtx {
    return buildCtx{
        Context: ctx,
        Dependencies: deps,
        BuilderState: builderState,
    }
}
```

This way all the `scbuildstmt` functions can look like:

```
func AlterTable(ctx BuildCtx, n *tree.AlterTable) {
```

I might even go one further and rename `BuildCtx` to `Context`.

**Review by ajwerner — COMMENTED** [review]
This is getting nice! I've left some minor suggestions which are getting into the realm of taste.

File: pkg/sql/schemachanger/scbuild/build.go
Comment (ajwerner) [reply pkg/sql/schemachanger/scbuild/build.go:39]: Okay, one more thought. I feel like the most compelling thing about keeping `ProcessStatement` in this package was that it clarified the error handling boundaries, namely, it clarified that all error propagation during statement building is panicked. I think we can and should maintain that. Consider:

```go
package scbuildstmt

// Process will process the provided statement. Errors are panicked.
func Process(ctx BuildCtx, n tree.Statement) {
	switch n := n.(type) {
	case *tree.AlterTable:
		scbuildstmts.AlterTable(ctx, bCtx, n)
	case *tree.CreateIndex:
		scbuildstmts.CreateIndex(ctx, bCtx, n)
	case *tree.DropDatabase:
		scbuildstmts.DropDatabase(ctx, bCtx, n)
	case *tree.DropSchema:
		scbuildstmts.DropSchema(ctx, bCtx, n)
	case *tree.DropSequence:
		scbuildstmts.DropSequence(ctx, bCtx, n)
	case *tree.DropTable:
		scbuildstmts.DropTable(ctx, bCtx, n)
	case *tree.DropType:
		scbuildstmts.DropType(ctx, bCtx, n)
	case *tree.DropView:
		scbuildstmts.DropView(ctx, bCtx, n)
	default:
		panic(scerrors.NotImplementedError(n))
	}
}
```

Then
```go
package scbuild

// Build ...
func Build(
	ctx context.Context, deps Dependencies, initial scpb.State, stmt tree.Statement,
) (_ scpb.State, err error) {
    	defer func() {
    	        switch r := recover().(type) {
    	        case nil: // no error
    	        case error:
    	            err = r
    	        default:
    	            err =  errors.AssertionFailedf(
    	                "unexpected error (%T) encountered while building schema change for %s: %s",
    	                 r, stmt, r,
    	             )
    	        }
	}()
	bs := newBuilderState(deps, initial, stmt)
	scbuildstmt.Process(makeBuildCtx(ctx, deps, bs), stmt)
	return *bs.output, nil
}
```

File: pkg/sql/schemachanger/scbuild/build.go
Comment (postamar) [reply pkg/sql/schemachanger/scbuild/build.go:58]: I... hadn't thought of that! Let's do it.

File: pkg/sql/schemachanger/scbuild/build.go
Comment (postamar) [reply pkg/sql/schemachanger/scbuild/build.go:39]: Embedding Context! I like it

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go
Comment (ajwerner) [inline pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go:103]: do we like that this is exposed up to the `scbuildstmt` package? Seems better to remove it and leave the process of asking predicate questions of nodes to the context interface. We can defer that to a subsequent change.

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go
Comment (ajwerner) [inline pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go:132]: supreme nit: put this at the top

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go
Comment (ajwerner) [inline pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go:113]: I really don't love that this is exposed. If you need it for the implementation of `buildCtx` expose a private interface in `scbuild` which extends `BuilderState`, or, in lieu of that, just have your one concrete impl of `buildCtx` take a concrete `*scbuild.builderState` and explicitly delegate the two methods here that you actually need to expose to this layer.

Comment (postamar) [reply pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go:113]: No, I really want to avoid exposing `builderState` internals. It's dangerous and begging to be abused. Also, since this mutates the state directly, I feel it belongs there. I understand your point that it feels stupid and I'll try to work out a better way of doing the same thing.

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go
Comment (postamar) [reply pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go:103]: I'm ambivalent about this and I'm totally open to revisiting this. I would like to defer this conversation until after future work involving element decomposition. What we _really want_ is not predicates on the builder state but something more powerful, some kind of _view_ on what there is so far, in terms of elements.

Comment (postamar) [reply pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go:103]: In other words this is going to change soon anyway so let's leave this as it is for the time being.

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go
Comment (ajwerner) [reply pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go:113]: How about:

```go
type internalBuilderState interface {
     scbuildstmt.BuilderState
     
     forEachNode(/* ... */)
     withNewSourceElementID() internalBuilderState
}

type buildCtx struct {
	scbuildstmts.Dependencies
	internalBuilderState
}
```

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go
Comment (ajwerner) [reply pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go:103]: Cool, we can leave it be for now. Regarding interfaces in terms of elements vs predicates, I only sort of agree. I agree that we want to decompose and that we want to use that decomposition to drive the implementation of the builder. Where I disagree is about how to expose that to the statement building business logic.

This is a bit pedantic, but I would not be mad about extremely high level functions which then all defer to a a data structure that represents a queryable view over the decomposition. I think such a technique will 1) aid in unit testing and code readability and 2) enforce a nice isolation between the complexity of processing statements and querying elements. 

tl;dr I agree with wanting to write exactly the code you're saying you want to write, but I want yet another layer of indirection to access it.

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go
Comment (postamar) [reply pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go:113]: We could do something like that but I think I came up with something nicer to replace `BuilderStateWithNewSourceElementID` which I'll push right now.

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go
Comment (postamar) [reply pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go]: What you describe is what I had in mind when I mentioned the need for a "view" earlier in this thread.

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go
Comment (ajwerner) [inline pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go:37]: You can unexport all of these now. I do get that there will be name collisions with existing functions, but I believe in your ability to sort that. I can't imagine anybody ever wanting to call one of these directly.

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go
Comment (ajwerner) [inline pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go:20]: Note that it panics on error.

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go
Comment (ajwerner) [inline pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go:28]: Note the expectation of panics for error handling on methods here which do not explicitly return errors.

**Review by ajwerner — APPROVED** [review]
LGTM

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go
Comment (postamar) [reply pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go:28]: Will do, good point.

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go
Comment (postamar) [reply pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go:20]: Will do, good point.

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go
Comment (postamar) [reply pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go:37]: Sure, but the package is internal, so it doesn't really matter, because nobody can call these directly if they wanted to. I'll leave it as is, if you don't mind.

Comment (ajwerner) [reply pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go:37]: 👍 I generally like the use `internal`, go's answer to `protected`.

**Comment by postamar** [comment]
Thanks for the review. I'm going to properly take a look at https://github.com/cockroachdb/cockroach/pull/72856 before merging this one. Either I'll wait until the other is merged, or I'll merge mine and help rebase the other (I don't want to be a jerk!).

**Review by fqazi — APPROVED** [review]
> Reviewed 18 of 58 files at r1, 9 of 44 files at r2, 23 of 25 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72854)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, and @postamar)

**Comment by fqazi** [comment]
@postamar You can go ahead first in terms of the merge since your stuff is ready. For my stuff I think the rebase will be mechanical so, I'll go through the exercise after (I don't think it will be too terrible).

**Comment by postamar** [comment]
Thanks Faizan. I suspected as much but I wasn't sure. You're the best judge of this in this matter so I defer to you. 

From what I can tell right now, your rebase will involve adding a few new methods to `BuildCtx`, I'll be very happy to help you with this any time you'd like.

**Pull request merged**
